pkg/pss/pss.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pss
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/pushsync"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
+	"github.com/ethersphere/bee/pkg/trojan"
+)
+
+// Pss is the top-level struct, which takes care of message sending
+type Pss struct {
+	pusher     pushsync.PushSyncer
+	tags       *tags.Tags
+	handlers   map[trojan.Topic]Handler
+	handlersMu sync.RWMutex
+	metrics    metrics
+	logger     logging.Logger
+}
+
+type Options struct {
+	Logger     logging.Logger
+	PushSyncer pushsync.PushSyncer
+	Tags       *tags.Tags
+}
+
+// NewPss inits the Pss struct with the storer
+func NewPss(o Options) *Pss {
+	return &Pss{
+		pusher:   o.PushSyncer,
+		tags:     o.Tags,
+		handlers: make(map[trojan.Topic]Handler),
+		metrics:  newMetrics(),
+		logger:   o.Logger,
+	}
+}
+
+// Handler defines code to be executed upon reception of a trojan message
+type Handler func(trojan.Message)
+
+// Send constructs a padded message with topic and payload,
+// wraps it in a trojan chunk such that one of the targets is a prefix of the chunk address
+// uses push-sync to deliver message
+func (p *Pss) Send(ctx context.Context, targets trojan.Targets, topic trojan.Topic, payload []byte) (*tags.Tag, error) {
+	p.metrics.TotalMessagesSentCounter.Inc()
+
+	//construct Trojan Chunk
+	m, err := trojan.NewMessage(topic, payload)
+	if err != nil {
+		return nil, err
+	}
+	var tc swarm.Chunk
+	tc, err = m.Wrap(targets)
+	if err != nil {
+		return nil, err
+	}
+
+	tag, err := p.tags.Create("pss-chunks-tag", 1, false)
```

Comment:
We should review this in the PSS refactoring, which I think is not going to be critical for global pinning as of now.

Reply:
@santicomp2014 if `PushChunkToClosest` returns without an error then the operation succeeded. all of the state you need and know about the chunk is already in this function. pushsync will not retry pushing the chunk on error so the tag does not bring any added value, i.e. even if you continue polling the tag in another context with the hopes of the chunk being eventually synced, the state of it wont change. this can be removed

Reply:
I see your point, so `PushChunkToClosest `  would block until message is sent, and if error arrives then it can be considered a failure and somebody up the stack can try a retry logic or not.

About tags, I thought the idea of the tag was to view the status of the chunk in its journey.
Is the tag not Sent and Synced currently?

This is an important question, I could remove tags altogether, but I would like @zelig and @jmozah opinion on this, before removing this from PSS.